### PR TITLE
Clean up queued batches on task failures in RapidsShuffleThreadedBlockIterator

### LINKED ISSUE #8043: [BUG] Host memory leak in SerializedBatchIterator
**Describe the bug**
While testing #7581 with NDS 3TB with GPU memory restricted to 6GB, I am seeing some leaked host memory buffers.
I see these with and without the fix in #8040 
```
Executor task launch worker for task 127.0 in stage 1653.0 (TID 71217) 23/04/05 18:34:02:230 INFO GpuParquetMultiFilePartitionReaderFactory: Using the coalesce multi-file Parquet reader, files: hdfs://rl-r7525-d32-u38.raplab.nvidia.com:9000/data/nds2.0/parquet_sf3k_decimal/store_sales/ss_sold_date_sk=2452221/part-00179-9dcfb50c-76b0-4dbf-882b-b60e7ad5b925.c000.snappy.parquet,hdfs://rl-r7525-d32-u38.raplab.nvidia.com:9000/data/nds2.0/parquet_sf3k_decimal/store_sales/ss_sold_date_sk=2452220/part-00073-9dcfb50c-76b0-4dbf-882b-b60e7ad5b925.c000.snappy.parquet,hdfs://rl-r7525-d32-u38.raplab.nvidia.com:9000/data/nds2.0/parquet_sf3k_decimal/store_sales/ss_sold_date_sk=2452244/part-00154-9dcfb50c-76b0-4dbf-882b-b60e7ad5b925.c000.snappy.parquet task attemptid: 71217
Cleaner Thread 23/04/05 18:34:02:478 ERROR HostMemoryBuffer: A HOST BUFFER WAS LEAKED (ID: 1897514 7f6758548f10)
Executor task launch worker for task 121.0 in stage 1653.0 (TID 71211) 23/04/05 18:34:02:480 INFO Executor: Finished task 121.0 in stage 1653.0 (TID 71211). 4570 bytes result sent to driver
dispatcher-Executor 23/04/05 18:34:02:481 INFO CoarseGrainedExecutorBackend: Got assigned task 71255
Executor task launch worker for task 15.0 in stage 1656.0 (TID 71255) 23/04/05 18:34:02:481 INFO Executor: Running task 15.0 in stage 1656.0 (TID 71255)
Executor task launch worker for task 15.0 in stage 1656.0 (TID 71255) 23/04/05 18:34:02:482 INFO TorrentBroadcast: Started reading broadcast variable 1541 with 1 pieces (estimated total size 4.0 MiB)
Executor task launch worker for task 15.0 in stage 1656.0 (TID 71255) 23/04/05 18:34:02:483 INFO MemoryStore: Block broadcast_1541_piece0 stored as bytes in memory (estimated size 10.3 KiB, free 8.2 GiB)
Executor task launch worker for task 15.0 in stage 1656.0 (TID 71255) 23/04/05 18:34:02:483 INFO TorrentBroadcast: Reading broadcast variable 1541 took 1 ms
Executor task launch worker for task 15.0 in stage 1656.0 (TID 71255) 23/04/05 18:34:02:484 INFO MemoryStore: Block broadcast_1541 stored as values in memory (estimated size 22.8 KiB, free 8.2 GiB)
Executor task launch worker for task 15.0 in stage 1656.0 (TID 71255) 23/04/05 18:34:02:485 INFO TorrentBroadcast: Started reading broadcast variable 1540 with 1 pieces (estimated total size 4.0 MiB)
Cleaner Thread 23/04/05 18:34:02:485 ERROR MemoryCleaner: Leaked host buffer (ID: 1897514): 2023-04-05 18:33:59.0349 UTC: INC
java.lang.Thread.getStackTrace(Thread.java:1559)
ai.rapids.cudf.MemoryCleaner$RefCountDebugItem.<init>(MemoryCleaner.java:333)
ai.rapids.cudf.MemoryCleaner$Cleaner.addRef(MemoryCleaner.java:91)
ai.rapids.cudf.MemoryBuffer.incRefCount(MemoryBuffer.java:275)
ai.rapids.cudf.MemoryBuffer.<init>(MemoryBuffer.java:117)
ai.rapids.cudf.HostMemoryBuffer.<init>(HostMemoryBuffer.java:196)
ai.rapids.cudf.HostMemoryBuffer.<init>(HostMemoryBuffer.java:192)
ai.rapids.cudf.HostMemoryBuffer.allocate(HostMemoryBuffer.java:144)
com.nvidia.spark.rapids.SerializedBatchIterator.$anonfun$tryReadNext$1(GpuColumnarBatchSerializer.scala:78)
com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
com.nvidia.spark.rapids.SerializedBatchIterator.withResource(GpuColumnarBatchSerializer.scala:34)
com.nvidia.spark.rapids.SerializedBatchIterator.tryReadNext(GpuColumnarBatchSerializer.scala:72)
com.nvidia.spark.rapids.SerializedBatchIterator.next(GpuColumnarBatchSerializer.scala:97)
org.apache.spark.sql.rapids.RapidsShuffleThreadedReaderBase$RapidsShuffleThreadedBlockIterator$BlockState.next(RapidsShuffleInternalManagerBase.scala:657)
org.apache.spark.sql.rapids.RapidsShuffleThreadedReaderBase$RapidsShuffleThreadedBlockIterator.$anonfun$deserializeTask$1(RapidsShuffleInternalManagerBase.scala:733)
java.util.concurrent.FutureTask.run(FutureTask.java:266)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
java.lang.Thread.run(Thread.java:748)
```
**Steps/Code to reproduce bug**
Run NDS at 3TB on an 8 node A100 cluster with GPU memory restricted to 6GB

### PR DESCRIPTION
**Pull request by jbrennan333**
Fixes #8043.

`RapidsShuffleThreadedBlockIterator` has a queue for batches from deserialized tasks that needs to be cleaned up on task failures.  I have not been able to reproduce #8043 with 23.06 builds, but I was able to reproduce with 23.04 by running NDS with reduced gpu memory (6g).  So I tested this by applying this patch to 23.04 and then verifying that I did not get any HOST BUFFER leaks.

### CODE REVIEW

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala [hunk 1]
```diff
@@ -636,6 +636,14 @@ abstract class RapidsShuffleThreadedReaderBase[K, C](
       null
     }
 
+    // Register a completion handler to close any queued cbs.
+    Option(TaskContext.get()).foreach {_.addTaskCompletionListener[Unit]( _ => {
```
Comment (jbrennan333) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala:640]: @abellina can I just use the `context` from `RapidsShuffleThreadedReaderBase` here?  Same applies for line 550.

**Comment by jbrennan333** [comment]
build

**Comment by jbrennan333** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala (hunk 1 above)
Comment (abellina) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala:640]: yes this is the same `TaskContext` as what is in `TaskContext.get`

Comment (jbrennan333) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala:640]: Thanks. I will update this.

**Review by abellina — APPROVED** [review]
Thanks for fixing this @jbrennan333!

Looking forward, perhaps something we can see later if we run into it: I wonder what happens in the case where the async operations here are very slow to process, and then the task completes or fails.

Do we need to cancel the futures? or anything else that is queued? and then poison pill things so that nothing else gets queued?

I do not know the right answer and it would be definitely follow on, but raising that because I realized I missed that as well when putting this code in.

**Comment by jbrennan333** [comment]
> Thanks for fixing this @jbrennan333!
> 
> Looking forward, perhaps something we can see later if we run into it: I wonder what happens in the case where the async operations here are very slow to process, and then the task completes or fails.
> 
> Do we need to cancel the futures? or anything else that is queued? and then poison pill things so that nothing else gets queued?
> 
> I do not know the right answer and it would be definitely follow on, but raising that because I realized I missed that as well when putting this code in.

I'm not sure either.  I could cancel and clear `futures` as part of this change if you want.  Seems like that would be safe.  But I'm not certain if we would also need a poison pill.   Let me know if you want me to add cleanup of futures here, or just file a follow-on for all of it.  The fact that I didn't get any host buffer leaks in my testing with this change suggests that nothing else was added to `queued` after the task completion handler ran, so at least in the test case I had, it didn't seem like any additional futures were running.

**Comment by abellina** [comment]
> 
> Thanks for fixing this @jbrennan333!
> 
> Looking forward, perhaps something we can see later if we run into it: I wonder what happens in the case where the async operations here are very slow to process, and then the task completes or fails.
> 
> Do we need to cancel the futures? or anything else that is queued? and then poison pill things so that nothing else gets queued?
> 
> I do not know the right answer and it would be definitely follow on, but raising that because I realized I missed that as well when putting this code in.
> 
> I'm not sure either. I could cancel and clear futures as part of this change if you want. Seems like that would be safe. But I'm not certain if we would also need a poison pill. Let me know if you want me to add cleanup of futures here, or just file a follow-on for all of it. The fact that I didn't get any host buffer leaks in my testing with this change suggests that nothing else was added to queued after the task completion handler ran, so at least in the test case I had, it didn't seem like any additional futures were running.

No definitely not for this PR. I would also punt until we have a test case for it for now. I mostly wanted to circle the thought in case I didn't miss something.

**Comment by jbrennan333** [comment]
build

**Pull request merged**